controller/src/controller/storage/kubernetes/fsadmin.py

```diff
@@ -0,0 +1,193 @@
+            objects.pod.metadata.name,
+            ns,
+            until_not={PodPhase.UNKNOWN, PodPhase.PENDING},
+            timeout=timeout,
+        )
+        self._start_time = current_datetime(microseconds=True)
+        return self._start_time
+
+    async def delete(self, objects: FSAdminObjects, timeout: Timeout) -> None:
+        """Delete the fsadmin instance.
+
+        Parameters
+        ----------
+        objects
+            Kubernetes objects making up the fsadmin environment.
+        timeout
+            Timeout on operation.
+
+        Raises
+        ------
+        KubernetesError
+            Raised if there is some failure in a Kubernetes API call.
+        TimeoutError
+            Raised if fsadmin objects are not deleted within provided timeout.
+        """
+        ns = self._metadata.namespace
+        pod = objects.pod
+        pvcs = objects.pvcs
+        self._start_time = None
+        await self._pod.delete(pod.metadata.name, ns, timeout, wait=True)
+        for pvc in pvcs:
+            await self._pvc.delete(pvc.metadata.name, ns, timeout)
+
+    async def is_fsadmin_ready(self, timeout: Timeout) -> bool:
+        """Check whether the fsadmin environment is ready for work.
+
+        Parameters
+        ----------
+        timeout
+            Timeout on operation.
+
+        Returns
+        -------
+        bool
+            Whether the fsadmin pod is ready.
+
+        Notes
+        -----
+        We check whether the pod exists, and if so, whether it is running.
+        We do not need to check the PVCs, because if they're necessary,
+        the pod can't be running unless it has them.
+        """
+        ns = self._metadata.namespace
+        existing_pod = await self._pod.read(self._config.pod_name, ns, timeout)
+        if existing_pod is None:  # No pod
+            return False
+        # Return whether it is running
+        return existing_pod.status.phase == PodPhase.RUNNING
+
+    async def get_start_time(
+        self, timeout: Timeout
+    ) -> datetime.datetime | None:
+        """Get time pod started, or None if it is not ready for work.
+
+        Parameters
+        ----------
+        timeout
+            How long to wait to query the fsadmin pod.
+
+        Returns
+        -------
+        datetime.datetime
+            The time the fsadmin pod went into ``Running`` phase.
+
+        Raises
+        ------
+        KubernetesError
+            Raised if there is some failure in a Kubernetes API call.
+        TimeoutError
+            Raised if fsadmin is not ready before the provided timeout expires.
+
+        Notes
+        -----
+        If the pod exists but is not yet running, or is in the process of
+        terminating, we report None. The start time is only valid for pods
+        which are ready to accept work.
+        """
+        if not await self.is_fsadmin_ready(timeout):
+            return None
+        return self._start_time
```

Comment:
You can replace both of these methods with one method that returns an `FSAdminStatus` if the pod is running or `None` if it's not.

Reply:
I think it's better to return an FSAdminStatus if it's running, and throw a ClientRequestError if it isn't explaining what the problem is.  Then I can catch that in the handler and turn it into an error return.  Otherwise I have to propagate what's really just the exception information (which comes down to, "there's no pod" or "the pod's not in Running state") through some additional method, which seems dumb.